MOVIES! I asked my girlfriend, brother, and three of my best friends to rate six movies on a scale of one to five: It ends with us, deadpool and wolverine, blink twice, despicable me 4, inside out 2, twisters. This notebook is an analysis of their ratings!

In [6]:
import pandas as pd

# Import and disply the data

df = pd.read_csv('/home/kobi/is362/Week 7 Assignment/Movies.csv')

display(df)

Name  It Ends with Us  Deadpool and Wolverine  Blink Twice  \
0  Nicole              4.0                     NaN          3.0   
1  Darian              5.0                     4.0          NaN   
2   Oscar              3.0                     5.0          2.0   
3  Adrian              4.0                     3.0          5.0   
4  Kismat              NaN                     5.0          4.0   

   Despicable Me 4  Inside Out 2  Twisters  
0              5.0             5       NaN  
1              3.0             5       4.0  
2              4.0             4       NaN  
3              3.0             4       2.0  
4              NaN             5       3.0

In [9]:
# Fill missing values
df_filled = df.fillna(df.mean(numeric_only=True))

# Average ratings for each user
user_avg = df_filled.set_index('Name').mean(axis=1)

# Average ratings for each movie
movie_avg = df_filled.drop(columns='Name').mean(axis=0)

print("Average ratings per user:")
print(user_avg)

print("\nAverage ratings per movie:")
print(movie_avg)

Average ratings per user:
Name
Nicole    4.041667
Darian    4.083333
Oscar     3.500000
Adrian    3.500000
Kismat    4.125000
dtype: float64

Average ratings per movie:
It Ends with Us           4.00
Deadpool and Wolverine    4.25
Blink Twice               3.50
Despicable Me 4           3.75
Inside Out 2              4.60
Twisters                  3.00
dtype: float64


Lets normalize this data according to the rating system of each individual! This can be done using a mathematical fomula: Xnew = (X-Xmin)/(Xmax-Xmin)

In [11]:
# Run the ratings through the mathematical fomula
df_normalized = df.set_index('Name').apply(lambda x: (x - x.min()) / (x.max() - x.min()), axis=1)

# Proceed with a similar paradigm as before calcualting average ratings for each user and movie.

user_avg_normalized = df_normalized.mean(axis=1)
movie_avg_normalized = df_normalized.mean(axis=0)
print("Average normalized ratings per user:")
print(user_avg_normalized)

print("\nAverage normalized ratings per movie:")
print(movie_avg_normalized)

Average normalized ratings per user:
Name
Nicole    0.625000
Darian    0.600000
Oscar     0.533333
Adrian    0.500000
Kismat    0.625000
dtype: float64

Average normalized ratings per movie:
It Ends with Us           0.625000
Deadpool and Wolverine    0.708333
Blink Twice               0.375000
Despicable Me 4           0.500000
Inside Out 2              0.866667
Twisters                  0.166667
dtype: float64


Let's now standardize this data according to the rating system of each individual. This can be done using a mathematical fomula: Xnew = (x - mu)/sigma

In [12]:
# Run the ratings through the mathematical fomula
df_standardized = df.set_index('Name').apply(lambda x: (x - x.mean()) / x.std(), axis=1)

# Proceed with a similar paradigm as before calcualting average ratings for each user and movie.
user_avg_standardized = df_standardized.mean(axis=1)
movie_avg_standardized = df_standardized.mean(axis=0)
print("Average standardized ratings per user:")
print(user_avg_standardized)

print("\nAverage standardized ratings per movie:")
print(movie_avg_standardized)

Average standardized ratings per user:
Name
Nicole    5.551115e-17
Darian   -1.887379e-16
Oscar    -8.881784e-17
Adrian    0.000000e+00
Kismat    5.551115e-17
dtype: float64

Average standardized ratings per movie:
It Ends with Us           0.161391
Deadpool and Wolverine    0.323863
Blink Twice              -0.384949
Despicable Me 4          -0.194208
Inside Out 2              0.670087
Twisters                 -0.991607
dtype: float64


It seems that Inside Out 2 was the most liked movie at an average rating of 4.6 and Twisters was the least liked at an average rating of 3.0. However, everyone has their own personal bias and standardizing/normalizing the data can help get a better idea of removing this personal bias. 

Normalizing the data helps adjust for people who rate more generously or more strictly (albeit, on a scale of 5, this may not be the best measure) to help make ratings comparable across people. After normalizing, the conclusion above regarding most liked and least liked movies did not change, however, it seems that Inside Out 2 was highly more favored at a rating of 0.8666666666 vs Twisters at a rating of 0.166666666. However, by normalizing, we lose insight into the actual distribution of ratings, which may be important for more complex analyses using other metrics. For example, if we were interested in how different age groups rate a particular movie, it may be beneficial to NOT normalize the data to investigate raw rating differences, both between age groups but also tracking variation within an age group. It may be the case that normalizing the data can help elucidate an age group difference in ratings for a particular movie, but the raw data may provide insight into bias and higher/lesser variation within a particular age group. 

Standardizing the data helps to focus on how much a rating deviates from someone's average, which may help provide insight into movies that someone enjoyed or didn't enjoy relative to their usual ratings. Again, it seems that Inside Out 2 was highly favored @ 0.67 while Twisters was not favored @ -0.99. The drawback of standardizing is that due to the nature of the mathematics, it may overly spread ratings from people who tend to rate within a narrow range.